File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.py
def solution(n, costs):
    # 비용순 정렬
    costs.sort(key = lambda x : x[2])
    # 비용, 깊이, 부모테이블
    answer = 0
    depth = 0
    parents = [i for i in range(n)]
    
    # 조상 찾기
    def find(node):
        if parents[node] == node:
            return node
        return find(parents[node])
    
    # 합치기
    def union(nodeA, nodeB):
        parentA = find(nodeA)
        parentB = find(nodeB)
        parents[parentB] = parentA
        
    # 크루스칼
    while depth < n-1:
        nodeA, nodeB, cost = costs.pop(0)
        if find(nodeA) != find(nodeB):
            union(nodeA, nodeB)
            answer += cost
            depth += 1
        
    
    return answer

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;

class Solution {
    
    int[] parent;
    
    public int solution(int n, int[][] costs) {

        // 부모
        parent = new int[n];
        for (int i = 0; i < n; i++) parent[i] = i;
        
        // 거리순으로 정렬하기
        Arrays.sort(costs, (a, b) -> a[2] - b[2]);
        
        int picked = 0;
        int answer = 0;
        for (int[] cost : costs) {
            // 다른 그룹이라면 union
            if (find(cost[0]) != find(cost[1])) {
                union(cost[0], cost[1]);
                picked++;
                answer += cost[2];
            }
            // 그룹 내 개수 체크
            if (picked == n-1) {
                return answer;
            }
        }
        
        return answer;
    }
    
    int find(int node) {
        if (parent[node] == node) return node;
        return find(parent[node]);
    }
    
    void union(int node1, int node2) {
        parent[find(node2)] = find(node1);
    }
}